Non constant size and offset in DWARF

tromey opened the pull request:
In Ada, a record type can have a non-constant size, and a field can appear at a non-constant bit offset in a record.

To support this, this patch changes DIType to record the size and offset using metadata, rather than plain integers.  In addition to a constant offset, both DIVariable and DIExpression are now supported here.

One thing of note in this patch is the choice of how exactly to represent a non-constant bit offset, with the difficulty being that DWARF 5 does not support this.  DWARF 3 did have a way to support a non-constant byte offset, combined with a constant bit offset within the byte, but this was deprecated in DWARF 4 and removed from DWARF 5.

This patch takes a simple approach: a DWARF extension allowing the use of an expression with DW_AT_data_bit_offset.  There is a corresponding DWARF issue, see https://dwarfstd.org/issues/250501.1.html.  The main reason for this approach is that it keeps API simplicity: just a single value is needed, rather than having separate data describing the byte offset and the bit within the byte.

tromey committed:
Make it simpler to add metadata to DIType

This patch changes code in DebugInfoMetadata.h to use symbolic
constants to refer to metadata slots.  This makes it a little simpler
(but still not entirely seamless) to add metadata to DIType.

tromey committed:
Use metadata for size and offset in DIType

This changes DIType to use metadata for the size and offset
information.  This patch doesn't have any functional changes yet; the
test is in the next patch.

tromey committed:
Change DWARF writer to handle dynamic offsets and sizes

This updates the DWARF writer to emit dynamic offsets and sizes for
members.

tromey committed:
Expose dynamic sizes and offsets via DIBuilder

This adds some new DIBuilder methods to expose the new dynamic size
and offset functionality.

tromey commented:
Note that I've written this patch as a series of smaller patches.  I know it'll all be squashed if it lands, but it may be simpler to review the pieces separately.

tromey commented:
I rebased this and fixed up the mlir problem.  I also changed how the overload resolution was done, so that a later patch that treats `nullptr` specially can be done without affecting existing call sites.

tromey commented on llvm/include/llvm/IR/DebugInfoMetadata.h:
```diff
@@ -966,20 +992,42 @@ class DIFixedPointType : public DIBasicType {
   LLVM_ABI static DIFixedPointType *
   getImpl(LLVMContext &Context, unsigned Tag, MDString *Name,
-          uint64_t SizeInBits, uint32_t AlignInBits, unsigned Encoding,
+          uint32_t SizeInBits, uint32_t AlignInBits, unsigned Encoding,
```
I didn't mean to change the type of `SizeInBits` here.  I don't think it is harmful as is, but I'll fix it and upload a new patch shortly.

tromey commented:
Ping.

dwblaikie commented on llvm/lib/AsmParser/LLParser.cpp:
```diff
@@ -4822,6 +4822,35 @@ struct MDSignedOrMDField : MDEitherFieldImpl<MDSignedField, MDField> {
+  }
+
+  Metadata *getValueAsMetadata(LLVMContext &Context) const {
+    if (isMDUnsignedField()) {
+      return ConstantAsMetadata::get(
+          ConstantInt::get(Type::getInt64Ty(Context), getMDUnsignedValue()));
+    } else if (isMDField()) {
+      return getMDFieldValue();
+    }
+    return nullptr;
```
LLVM's style avoids "else after return", favouring code like this:
```
if (...)
  return;
if (...)
  return;
return;
```
(& omitting `{}` from single-line blocks)

tromey replied:
I fixed this here and I fixed the `{}` issue everywhere I could find it in the patch.

dwblaikie commented on llvm/lib/Bitcode/Reader/MetadataLoader.cpp:
```diff
@@ -1525,15 +1525,24 @@ Error MetadataLoader::MetadataLoaderImpl::parseOneMetadata(
                                 : DINode::FlagZero;
     uint32_t NumExtraInhabitants = (Record.size() > 7) ? Record[7] : 0;
 
+    Metadata *SizeInBits;
+    if (SizeIsMetadata) {
+      SizeInBits = getMDOrNull(Record[3]);
+    } else {
+      SizeInBits = ConstantAsMetadata::get(
+          ConstantInt::get(Type::getInt64Ty(Context), Record[3]));
+    }
```
Could use a conditional operator here (& similar places) - but I don't feel super strongly about it.

This appears a few times - might be worth a helper function?

tromey replied:
I added a helper here and it really cleans things up nicely.

dwblaikie commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -834,8 +834,8 @@ llvm::DIType *CGDebugInfo::CreateType(const BuiltinType *BT) {
     ObjTy = DBuilder.createStructType(TheCU, "objc_object", TheCU->getFile(), 0,
-                                      0, 0, llvm::DINode::FlagZero, nullptr,
-                                      llvm::DINodeArray());
+                                      (uint64_t)0, 0, llvm::DINode::FlagZero,
```
Hrm, not sure how I feel about it - but not strongly enough to push for something else.

But when this is being committed, could commit this clang change separately before the rest of the change, since it's not dependent. (but it's also a really small change/not a big deal to go in/get reverted if needed/etc)

tromey replied:
Agreed elsewhere in the thread to leave this as-is.

tromey commented:
I've rebased this and addressed the review comments.  I am leaving the one sub-thread open since I wasn't sure if I should send a separate PR (I don't have push rights so I can't land something separately on my own) -- just let me know if you'd like that, it's no trouble.  Thanks.

dwblaikie commented on llvm/unittests/IR/DebugInfoTest.cpp:
```diff
@@ -1301,4 +1301,33 @@ TEST(DIBuilder, CompositeTypes) {
 
+TEST(DIBuilder, DynamicOffsetAndSize) {
+  LLVMContext Ctx;
+  std::unique_ptr<Module> M(new Module("MyModule", Ctx));
```
Prefer `std::make_unique` and optionally omit the type on the LHS since it'll be clear from the RHS, like this:
```
auto M = std::make_unique<Module>("MyModule", Ctx);
```

tromey replied:
Fixed in the next version.

dwblaikie reviewed (approved):
Yeah, let's just go with this as a single patch.

Looking closer, in retrospect/for next time, perhaps this could've been a few smaller patches (I guess the strtuct, member, and bit field member are all separate except that they depend on the new `MDUnsignedOrMDField` - either that could've been included in one of the patches that needed it, and the other two patches dependent on that one arbitrarily chosen - or if that could've been added/tested independently, then all 3 functional patches independently depend on it - (plus the already discussed clang change that could've been done separately - but yeah, annoying to have a dependent patch for such a small change, so probably fine to keep it in in situations like this then just send a tiny PR real quick to go in before the rest once approved)))

vvereschaka commented:
Hi @tromey,

> LLVM Buildbot has detected a new failure on builder `llvm-nvptx64-nvidia-win` running on `as-builder-8` while building `clang,llvm` at step 7 "test-build-unified-tree-check-llvm".

there is the same problem with the `dynamic-bitfield.ll` test on [llvm-nvptx-nvidia-win](https://lab.llvm.org/buildbot/#/builders/54) builder
https://lab.llvm.org/buildbot/#/builders/54/builds/10317

dwblaikie commented:
> Hi @tromey,
> 
> > LLVM Buildbot has detected a new failure on builder `llvm-nvptx64-nvidia-win` running on `as-builder-8` while building `clang,llvm` at step 7 "test-build-unified-tree-check-llvm".
> 
> there is the same problem with the `dynamic-bitfield.ll` test on [llvm-nvptx-nvidia-win](https://lab.llvm.org/buildbot/#/builders/54) builder https://lab.llvm.org/buildbot/#/builders/54/builds/10317

https://github.com/llvm/llvm-project/pull/145797

If anyone wants to stamp that & I'll commit it once the presubmit checks run - hopefully that's adequate to address the nvptx issue - just at a very rough guess.

brendandahl commented:
It looks like this PR is causing [failures in emscripten's LTO builds](https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket/8711055186216400577/+/u/Emscripten_testsuite__ThinLTO_/stdout). I haven't looked into it yet, but if you have any thoughts on what may be the issue it would be helpful.
```
 #0 0x0000577663876d28 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/b/s/w/ir/x/w/install/bin/wasm-ld+0x1b8cd28)
 #1 0x00005776638747b5 llvm::sys::RunSignalHandlers() (/b/s/w/ir/x/w/install/bin/wasm-ld+0x1b8a7b5)
 #2 0x0000577663877576 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x0000742536842520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007425368969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007425368969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007425368969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x0000742536842476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007425368287f3 abort ./stdlib/abort.c:81:7
 #9 0x000074253682871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x0000742536839e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x00005776662f46a4 (/b/s/w/ir/x/w/install/bin/wasm-ld+0x460a6a4)
#12 0x00005776662eb044 llvm::MetadataLoader::MetadataLoaderImpl::resolveForwardRefsAndPlaceholders((anonymous namespace)::(anonymous namespace)::PlaceholderQueue&) MetadataLoader.cpp:0:0
#13 0x00005776662f3a48 llvm::MetadataLoader::MetadataLoaderImpl::getMetadataFwdRefOrLoad(unsigned int) MetadataLoader.cpp:0:0
#14 0x00005776662ea465 llvm::MetadataLoader::MetadataLoaderImpl::parseGlobalObjectAttachment(llvm::GlobalObject&, llvm::ArrayRef<unsigned long>) (/b/s/w/ir/x/w/install/bin/wasm-ld+0x4600465)
#15 0x00005776662ea02e llvm::MetadataLoader::MetadataLoaderImpl::loadGlobalDeclAttachments() (/b/s/w/ir/x/w/install/bin/wasm-ld+0x460002e)
#16 0x00005776662ea875 llvm::MetadataLoader::MetadataLoaderImpl::parseMetadata(bool) (/b/s/w/ir/x/w/install/bin/wasm-ld+0x4600875)
#17 0x00005776662f4fbc llvm::MetadataLoader::parseMetadata(bool) (/b/s/w/ir/x/w/install/bin/wasm-ld+0x460afbc)
#18 0x00005776662bd30b (anonymous namespace)::BitcodeReader::materializeMetadata() BitcodeReader.cpp:0:0
#19 0x0000577666518e76 llvm::Module::materializeMetadata() (/b/s/w/ir/x/w/install/bin/wasm-ld+0x482ee76)
#20 0x0000577664f35910 llvm::FunctionImporter::importFunctions(llvm::Module&, llvm::FunctionImporter::ImportMapTy const&) (/b/s/w/ir/x/w/install/bin/wasm-ld+0x324b910)
#21 0x00005776642b69f7 llvm::lto::thinBackend(llvm::lto::Config const&, unsigned int, std::__2::function<llvm::Expected<std::__2::unique_ptr<llvm::CachedFileStream, std::__2::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, llvm::Module&, llvm::ModuleSummaryIndex const&, llvm::FunctionImporter::ImportMapTy const&, llvm::DenseMap<unsigned long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long, void>, llvm::detail::DenseMapPair<unsigned long, llvm::GlobalValueSummary*>> const&, llvm::MapVector<llvm::StringRef, llvm::BitcodeModule, llvm::DenseMap<llvm::StringRef, unsigned int, llvm::DenseMapInfo<llvm::StringRef, void>, llvm::detail::DenseMapPair<llvm::StringRef, unsigned int>>, llvm::SmallVector<std::__2::pair<llvm::StringRef, llvm::BitcodeModule>, 0u>>*, bool, std::__2::function<llvm::Expected<std::__2::unique_ptr<llvm::CachedFileStream, std::__2::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, std::__2::vector<unsigned char, std::__2::allocator<unsigned char>> const&) (/b/s/w/ir/x/w/install/bin/wasm-ld+0x25cc9f7)
#22 0x00005776642a82b2 (anonymous namespace)::InProcessThinBackend::runThinLTOBackendThread(std::__2::function<llvm::Expected<std::__2::unique_ptr<llvm::CachedFileStream, std::__2::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, llvm::FileCache, unsigned int, llvm::BitcodeModule, llvm::ModuleSummaryIndex&, llvm::FunctionImporter::ImportMapTy const&, llvm::DenseSet<llvm::ValueInfo, llvm::DenseMapInfo<llvm::ValueInfo, void>> const&, std::__2::map<unsigned long, llvm::GlobalValue::LinkageTypes, std::__2::less<unsigned long>, std::__2::allocator<std::__2::pair<unsigned long const, llvm::GlobalValue::LinkageTypes>>> const&, llvm::DenseMap<unsigned long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long, void>, llvm::detail::DenseMapPair<unsigned long, llvm::GlobalValueSummary*>> const&, llvm::MapVector<llvm::StringRef, llvm::BitcodeModule, llvm::DenseMap<llvm::StringRef, unsigned int, llvm::DenseMapInfo<llvm::StringRef, void>, llvm::detail::DenseMapPair<llvm::StringRef, unsigned int>>, llvm::SmallVector<std::__2::pair<llvm::StringRef, llvm::BitcodeModule>, 0u>>&)::'lambda'(std::__2::function<llvm::Expected<std::__2::unique_ptr<llvm::CachedFileStream, std::__2::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>)::operator()(std::__2::function<llvm::Expected<std::__2::unique_ptr<llvm::CachedFileStream, std::__2::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>) const LTO.cpp:0:0
#23 0x00005776642a7523 (anonymous namespace)::InProcessThinBackend::runThinLTOBackendThread(std::__2::function<llvm::Expected<std::__2::unique_ptr<llvm::CachedFileStream, std::__2::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, llvm::FileCache, unsigned int, llvm::BitcodeModule, llvm::ModuleSummaryIndex&, llvm::FunctionImporter::ImportMapTy const&, llvm::DenseSet<llvm::ValueInfo, llvm::DenseMapInfo<llvm::ValueInfo, void>> const&, std::__2::map<unsigned long, llvm::GlobalValue::LinkageTypes, std::__2::less<unsigned long>, std::__2::allocator<std::__2::pair<unsigned long const, llvm::GlobalValue::LinkageTypes>>> const&, llvm::DenseMap<unsigned long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long, void>, llvm::detail::DenseMapPair<unsigned long, llvm::GlobalValueSummary*>> const&, llvm::MapVector<llvm::StringRef, llvm::BitcodeModule, llvm::DenseMap<llvm::StringRef, unsigned int, llvm::DenseMapInfo<llvm::StringRef, void>, llvm::detail::DenseMapPair<llvm::StringRef, unsigned int>>, llvm::SmallVector<std::__2::pair<llvm::StringRef, llvm::BitcodeModule>, 0u>>&) LTO.cpp:0:0
#24 0x00005776642a7e44 void std::__2::__function::__policy_func<void ()>::__call_func[abi:nn210000]<std::__2::__bind<(anonymous namespace)::InProcessThinBackend::start(unsigned int, llvm::BitcodeModule, llvm::FunctionImporter::ImportMapTy const&, llvm::DenseSet<llvm::ValueInfo, llvm::DenseMapInfo<llvm::ValueInfo, void>> const&, std::__2::map<unsigned long, llvm::GlobalValue::LinkageTypes, std::__2::less<unsigned long>, std::__2::allocator<std::__2::pair<unsigned long const, llvm::GlobalValue::LinkageTypes>>> const&, llvm::MapVector<llvm::StringRef, llvm::BitcodeModule, llvm::DenseMap<llvm::StringRef, unsigned int, llvm::DenseMapInfo<llvm::StringRef, void>, llvm::detail::DenseMapPair<llvm::StringRef, unsigned int>>, llvm::SmallVector<std::__2::pair<llvm::StringRef, llvm::BitcodeModule>, 0u>>&)::'lambda'(llvm::BitcodeModule, llvm::ModuleSummaryIndex&, llvm::FunctionImporter::ImportMapTy const&, llvm::DenseSet<llvm::ValueInfo, llvm::DenseMapInfo<llvm::ValueInfo, void>> const&, std::__2::map<unsigned long, llvm::GlobalValue::LinkageTypes, std::__2::less<unsigned long>, std::__2::allocator<std::__2::pair<unsigned long const, llvm::GlobalValue::LinkageTypes>>> const&, llvm::DenseMap<unsigned long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long, void>, llvm::detail::DenseMapPair<unsigned long, llvm::GlobalValueSummary*>> const&, llvm::MapVector<llvm::StringRef, llvm::BitcodeModule, llvm::DenseMap<llvm::StringRef, unsigned int, llvm::DenseMapInfo<llvm::StringRef, void>, llvm::detail::DenseMapPair<llvm::StringRef, unsigned int>>, llvm::SmallVector<std::__2::pair<llvm::StringRef, llvm::BitcodeModule>, 0u>>&), llvm::BitcodeModule&, std::__2::reference_wrapper<llvm::ModuleSummaryIndex>, std::__2::reference_wrapper<llvm::FunctionImporter::ImportMapTy const>, std::__2::reference_wrapper<llvm::DenseSet<llvm::ValueInfo, llvm::DenseMapInfo<llvm::ValueInfo, void>> const>, std::__2::reference_wrapper<std::__2::map<unsigned long, llvm::GlobalValue::LinkageTypes, std::__2::less<unsigned long>, std::__2::allocator<std::__2::pair<unsigned long const, llvm::GlobalValue::LinkageTypes>>> const>, std::__2::reference_wrapper<llvm::DenseMap<unsigned long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long, void>, llvm::detail::DenseMapPair<unsigned long, llvm::GlobalValueSummary*>> const>, std::__2::reference_wrapper<llvm::MapVector<llvm::StringRef, llvm::BitcodeModule, llvm::DenseMap<llvm::StringRef, unsigned int, llvm::DenseMapInfo<llvm::StringRef, void>, llvm::detail::DenseMapPair<llvm::StringRef, unsigned int>>, llvm::SmallVector<std::__2::pair<llvm::StringRef, llvm::BitcodeModule>, 0u>>>>>(std::__2::__function::__policy_storage const*) LTO.cpp:0:0
#25 0x0000577663bf898e std::__2::__deferred_assoc_state<void, std::__2::__async_func<std::__2::function<void ()>>>::__execute() BalancedPartitioning.cpp:0:0
#26 0x0000577663bff4ed llvm::StdThreadPool::processTasks(llvm::ThreadPoolTaskGroup*) (/b/s/w/ir/x/w/install/bin/wasm-ld+0x1f154ed)
#27 0x0000577663c00f5e void* llvm::thread::ThreadProxy<std::__2::tuple<llvm::StdThreadPool::grow(int)::$_0>>(void*) ThreadPool.cpp:0:0
#28 0x0000742536894ac3 start_thread ./nptl/pthread_create.c:442:8
#29 0x0000742536926850 ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:83:0
```

anemet commented:
We're also seeing LTO failures with this: 

https://green.lab.llvm.org/job/llvm.org/job/clang-stage2-Rthinlto/900/

I think that this should be reverted.  I also have a small reproducer but unfortunately it requires the Apple linker.

tromey commented:
I have a fix for the crash that I'll send momentarily.
